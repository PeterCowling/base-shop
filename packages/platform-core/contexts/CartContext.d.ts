import type { CartState, SKU } from "@types";
import { ReactNode } from "react";
type Action = {
    type: "add";
    sku: SKU;
    size?: string;
} | {
    type: "remove";
    id: SKU["id"];
} | {
    type: "setQty";
    id: SKU["id"];
    qty: number;
};
export declare function CartProvider({ children }: {
    children: ReactNode;
}): import("react/jsx-runtime").JSX.Element;
export declare function useCart(): [CartState, import("react").Dispatch<Action>];
export {};
