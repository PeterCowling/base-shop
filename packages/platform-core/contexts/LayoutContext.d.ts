import { ReactNode } from "react";
export interface LayoutContextValue {
    isMobileNavOpen: boolean;
    breadcrumbs: string[];
    toggleNav: () => void;
}
export declare function LayoutProvider({ children }: {
    children: ReactNode;
}): import("react").JSX.Element;
export declare function useLayout(): LayoutContextValue;
