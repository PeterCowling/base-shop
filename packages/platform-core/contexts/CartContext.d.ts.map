{"version":3,"file":"CartContext.d.ts","sourceRoot":"","sources":["CartContext.tsx"],"names":[],"mappings":"AAOA,OAAO,KAAK,EAAE,GAAG,EAAE,MAAM,gBAAgB,CAAC;AAC1C,OAAO,EAEL,SAAS,EAIV,MAAM,OAAO,CAAC;AAIf,KAAK,QAAQ,GAAG;IAAE,GAAG,EAAE,GAAG,CAAC;IAAC,GAAG,EAAE,MAAM,CAAA;CAAE,CAAC;AAC1C,MAAM,MAAM,SAAS,GAAG,MAAM,CAAC,GAAG,CAAC,IAAI,CAAC,EAAE,QAAQ,CAAC,CAAC;AAEpD,KAAK,MAAM,GACP;IAAE,IAAI,EAAE,KAAK,CAAC;IAAC,GAAG,EAAE,GAAG,CAAA;CAAE,GACzB;IAAE,IAAI,EAAE,QAAQ,CAAC;IAAC,EAAE,EAAE,GAAG,CAAC,IAAI,CAAC,CAAA;CAAE,GACjC;IAAE,IAAI,EAAE,QAAQ,CAAC;IAAC,EAAE,EAAE,GAAG,CAAC,IAAI,CAAC,CAAC;IAAC,GAAG,EAAE,MAAM,CAAA;CAAE,CAAC;AA4DnD,wBAAgB,YAAY,CAAC,EAAE,QAAQ,EAAE,EAAE;IAAE,QAAQ,EAAE,SAAS,CAAA;CAAE,+BAUjE;AAED,wBAAgB,OAAO,kDAItB"}