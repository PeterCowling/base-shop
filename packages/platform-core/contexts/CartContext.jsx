"use client";

import { createContext, useContext, useEffect, useReducer } from "react";
import {
  asSetCookieHeader,
  CART_COOKIE,
  encodeCartCookie,
} from "../cartCookie";

/* ---------- reducer ---------- */
function reducer(state, action) {
  switch (action.type) {
    case "add": {
      const line = state[action.sku.id];
      return {
        ...state,
        [action.sku.id]: {
          sku: action.sku,
          qty: (line?.qty ?? 0) + 1,
          size: action.size ?? line?.size,
        },
      };
    }
    case "remove": {
      const next = { ...state };
      delete next[action.id];
      return next;
    }
    case "setQty": {
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          qty: Math.max(1, action.qty),
        },
      };
    }
    default:
      return state;
  }
}
/* ---------- persistence helpers ---------- */
const LS_KEY = CART_COOKIE;
function readInitial() {
  if (typeof window === "undefined") return {};
  try {
    const decoded = decodeURIComponent(raw);
    const parsed = JSON.parse(decoded);
    return cartStateSchema.parse(parsed);
  } catch (err) {
    console.warn("Invalid cart state in localStorage", err);
    return {};
    return {};
  }
}
function persist(state) {
  if (typeof window !== "undefined") {
    const encoded = encodeCartCookie(state);
    /* localStorage for client rehydration */
    localStorage.setItem(LS_KEY, encoded);
    /* cookie for server-side pages (e.g. /checkout) */
    document.cookie = asSetCookieHeader(encoded);
  }
}
/* ---------- context ---------- */
const CartContext = createContext(undefined);
export function CartProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, {}, readInitial);
  useEffect(() => persist(state), [state]);
  return (
    <CartContext.Provider value={[state, dispatch]}>
      {children}
    </CartContext.Provider>
  );
}
export function useCart() {
  const ctx = useContext(CartContext);
  if (!ctx) throw new Error("useCart must be inside CartProvider");
  return ctx;
}
