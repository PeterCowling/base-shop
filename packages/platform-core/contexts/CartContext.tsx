/* packages/platform-core/contexts/CartContext.tsx */
"use client";

import type { CartState, SKU } from "@types";
import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useReducer,
} from "react";
import {
  asSetCookieHeader,
  CART_COOKIE,
  encodeCartCookie,
} from "../cartCookie";

/* ---------- types ---------- */

type Action =
  | { type: "add"; sku: SKU }
  | { type: "remove"; id: SKU["id"] }
  | { type: "setQty"; id: SKU["id"]; qty: number };

/* ---------- reducer ---------- */

function reducer(state: CartState, action: Action): CartState {
  switch (action.type) {
    case "add": {
      const line = state[action.sku.id];
      return {
        ...state,
        [action.sku.id]: { sku: action.sku, qty: (line?.qty ?? 0) + 1 },
      };
    }

    case "remove": {
      /** remove without creating an unused binding */
      const next = { ...state };
      delete next[action.id];
      return next;
    }

    case "setQty": {
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          qty: Math.max(1, action.qty),
        },
      };
    }

    default:
      return state;
  }
}

/* ---------- persistence helpers ---------- */

const LS_KEY = CART_COOKIE;

function readInitial(): CartState {
  if (typeof window === "undefined") return {};
  try {
    return JSON.parse(
      decodeURIComponent(localStorage.getItem(LS_KEY) || "{}")
    ) as CartState;
  } catch {
    return {};
  }
}

function persist(state: CartState) {
  if (typeof window !== "undefined") {
    const encoded = encodeCartCookie(state);

    // localStorage for client re-hydration
    localStorage.setItem(LS_KEY, encoded);

    // cookie for SSR routes (e.g. /checkout)
    document.cookie = asSetCookieHeader(encoded);
  }
}

/* ---------- context ---------- */

const CartContext = createContext<
  [CartState, React.Dispatch<Action>] | undefined
>(undefined);

export function CartProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(reducer, {}, readInitial);

  useEffect(() => persist(state), [state]);

  return (
    <CartContext.Provider value={[state, dispatch]}>
      {children}
    </CartContext.Provider>
  );
}

export function useCart() {
  const ctx = useContext(CartContext);
  if (!ctx) throw new Error("useCart must be inside CartProvider");
  return ctx;
}
