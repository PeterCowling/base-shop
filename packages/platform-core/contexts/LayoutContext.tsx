"use client";

import { usePathname } from "next/navigation";
import { createContext, ReactNode, useContext, useState } from "react";

export interface LayoutContextValue {
  isMobileNavOpen: boolean;
  breadcrumbs: string[];
  toggleNav: () => void;
}

const LayoutContext = createContext<LayoutContextValue | undefined>(undefined);

export function LayoutProvider({ children }: { children: ReactNode }) {
  const [isMobileNavOpen, setIsMobileNavOpen] = useState(false);
  const pathname = usePathname();
  const breadcrumbs = pathname ? pathname.split("/").filter(Boolean) : [];

  const toggleNav = () => setIsMobileNavOpen((v) => !v);

  return (
    <LayoutContext.Provider value={{ isMobileNavOpen, breadcrumbs, toggleNav }}>
      {children}
    </LayoutContext.Provider>
  );
}

export function useLayout() {
  const ctx = useContext(LayoutContext);
  if (!ctx) throw new Error("useLayout must be inside LayoutProvider");
  return ctx;
}
