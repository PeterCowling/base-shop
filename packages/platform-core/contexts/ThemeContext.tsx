"use client";

import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";

export type Theme = "base" | "dark" | "brandx";

interface ThemeContextValue {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}

const ThemeContext = createContext<ThemeContextValue | undefined>(undefined);

export function ThemeProvider({ children }: { children: ReactNode }) {
  const [theme, setTheme] = useState<Theme>("base");

  useEffect(() => {
    const root = document.documentElement;
    root.classList.remove("theme-dark", "theme-brandx");
    if (theme === "dark") root.classList.add("theme-dark");
    if (theme === "brandx") root.classList.add("theme-brandx");
  }, [theme]);

  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
}

export function useTheme() {
  const ctx = useContext(ThemeContext);
  if (!ctx) throw new Error("useTheme must be inside ThemeProvider");
  return ctx;
}
