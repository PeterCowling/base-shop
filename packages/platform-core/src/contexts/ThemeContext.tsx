"use client";

import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useLayoutEffect,
  useState,
} from "react";

export type Theme = "base" | "dark" | "brandx" | "system";

interface ThemeContextValue {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}

const ThemeContext = createContext<ThemeContextValue | undefined>(undefined);

export function getSavedTheme(): Theme | null {
  if (typeof window === "undefined") return null;
  return window.localStorage.getItem("theme") as Theme | null;
}

function getSystemTheme(): Theme {
  return window.matchMedia("(prefers-color-scheme: dark)").matches
    ? "dark"
    : "base";
}

function getInitialTheme(): Theme {
  const stored = getSavedTheme();
  if (stored) return stored;
  return "system";
}

export function ThemeProvider({ children }: { children: ReactNode }) {
  const [theme, setTheme] = useState<Theme>(getInitialTheme);
  const [systemTheme, setSystemTheme] = useState<Theme>(getSystemTheme());

  useLayoutEffect(() => {
    const root = document.documentElement;
    root.classList.remove("theme-dark", "theme-brandx");

    const applied = theme === "system" ? systemTheme : theme;
    if (applied === "dark") root.classList.add("theme-dark");
    if (applied === "brandx") root.classList.add("theme-brandx");
    window.localStorage.setItem("theme", theme);
  }, [theme, systemTheme]);

  useEffect(() => {
    if (theme !== "system") return;
    const mq = window.matchMedia("(prefers-color-scheme: dark)");
    const update = (e: MediaQueryList | MediaQueryListEvent) =>
      setSystemTheme(e.matches ? "dark" : "base");
    update(mq);
    mq.addEventListener("change", update);
    return () => mq.removeEventListener("change", update);
  }, [theme]);

  useEffect(() => {
    const handler = (e: StorageEvent) => {
      if (e.key === "theme" && e.newValue) {
        setTheme(e.newValue as Theme);
      }
    };
    window.addEventListener("storage", handler);
    return () => {
      window.removeEventListener("storage", handler);
    };
  }, [setTheme]);

  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
}

export function useTheme() {
  const ctx = useContext(ThemeContext);
  if (!ctx) throw new Error("useTheme must be inside ThemeProvider");
  return ctx;
}
