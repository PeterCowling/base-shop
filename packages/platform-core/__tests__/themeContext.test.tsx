import { fireEvent, render } from "@testing-library/react";
import { ThemeProvider, useTheme } from "../contexts/ThemeContext";

function Buttons() {
  const { setTheme } = useTheme();
  return (
    <div>
      <button onClick={() => setTheme("dark")}>dark</button>
      <button onClick={() => setTheme("brandx")}>brandx</button>
    </div>
  );
}

describe("ThemeContext", () => {
  afterEach(() => {
    document.documentElement.className = "";
  });

  it("applies classes to <html> element", () => {
    const { getByText } = render(
      <ThemeProvider>
        <Buttons />
      </ThemeProvider>
    );

    const html = document.documentElement;
    expect(html.className).toBe("");
    fireEvent.click(getByText("dark"));
    expect(html.classList.contains("theme-dark")).toBe(true);
    fireEvent.click(getByText("brandx"));
    expect(html.classList.contains("theme-dark")).toBe(false);
    expect(html.classList.contains("theme-brandx")).toBe(true);
  });

  it("matches snapshot on initial render", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <div>Hello</div>
      </ThemeProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
