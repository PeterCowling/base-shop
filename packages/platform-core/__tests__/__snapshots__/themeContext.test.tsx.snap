// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeContext matches snapshot on initial render 1`] = `
<DocumentFragment>
  <div>
    Hello
  </div>
</DocumentFragment>
`;